### DISCUSSION TITLE
Extending the Subcircuit class

### DISCUSSION THREAD

**Discussion by jaminSch** [SpiceSharp/SpiceSharp#138] [Q&A]
Hello :)
I am currently trying to create a Switch out of two resistors and a sampler (btw. the sampler works fine!).
I wanted to create this Switch as an extension of the Subcircuit class but the behaviors are no longer recognized in the process of creating them. In my understanding the behaviors are created depending on the name of the component. Does this mean i have to rewrite all behaviors for my Switch or is there a way to use the existing Subcircuit class? I am not 100% confident about how the behaviors are working therefore i would really prefere using the existing class.
Thanks for your help and have a nice weekend!

**Comment by svenboulanger** [comment]
Hi @jaminSch ,

Happy to hear the sampler is working out :-).

> I wanted to create this Switch as an extension of the Subcircuit class but the behaviors are no longer recognized in the process of creating them.

The `Subcircuit` is kind of a special component that allows you to define a small circuit, and instance it multiple times with different connections. I'm afraid I'm not sure what you mean by 'no longer recognized'. Could you elaborate?

---

> In my understanding the behaviors are created depending on the name of the component.

[EDIT] I just realized that you mean that behaviors take over the name of their component. You can skip the next paragraph (I'll leave it for more information).

Behaviors should be created by the component itself. However, in Spice# v3 a technique called "inversion of control" is used. This means that the attribute `BehaviorForAttribute` allows the behavior to decide for which component it should be used, and a central container class (here `DI`) then searches for behaviors with this attribute and passes them to any entity that asks for its behaviors.
(I should note that the IOC concept is going to be removed again for the next version for multiple reasons)

---

> Does this mean I have to rewrite all behaviors for my Switch or is there a way to use the existing Subcircuit class?

[EDIT] After writing this I realized that you probably didn't really ask for how to use them. You can skip this part (I'll leave it as information).

In order to use the subcircuit class, you first create a subcircuit definition that represents it:

```c#
// Define the subcircuit
var subckt = new SubcircuitDefinition(new Circuit(
    new Resistor("R1", "a", "b", 1e3),
    new Resistor("R2", "b", "0", 1e3)),
    "a", "b");
```

Once you have your subcircuit definition, you can instance it using the `Subcircuit` component:

```c#
// Define the circuit
var ckt = new Circuit(
    new VoltageSource("V1", "in", "0", 5.0),
    new Subcircuit("X1", subckt).Connect("in", "out"));
```

---

> I am currently trying to create a Switch out of two resistors and a sampler (btw. the sampler works fine!).

I am not a 100% sure here, but I think you might be attempting to access these resistors from the sampler. This is actually not that easy to do, because the subcircuit will shield the behaviors inside the subcircuit from the outside to avoid behaviors with duplicate names in the simulation.
If that is the case, it might be more interesting to create a dedicated method for it, that just adds these behaviors flatly.

```c#
void AddSwitch(Circuit ckt, string name, string a, string b)
{
    // Create a node name that won't be used anywhere else
    var c = "c_" + name;

    // Create a circuit with names that won't be used anywhere else
    ckt.Add(new Resistor(name + ".R1", a, c, 1e3));
    ckt.Add(new Resistor(name + ".R2", c, b, 1e3));
    // ...
}
```

Doing this avoids any of these issues.

That being said, I think there is a case to be made for making subcircuits easier in use. The original use for subcircuits was to allow parallel computing, but maybe it is worth investing some more time in them for usability.

I hope this helps!

**Reply by jaminSch** [reply]
Thank you for your fast and detailed reply! Really appreciate it!
I think it would be good if i explained what i want to do a bit more:
I try to simulate a switch by using two resistors one with a really high and one with a really low resistance.
![MySwitch_prop](https://user-images.githubusercontent.com/57145245/107016374-2947d800-679e-11eb-9b92-8482d5f164ed.PNG)

This Switch should periodically change its status by changing the resistance values of the two resistors. 
Since the Resistor does not have influence on the timesteps inside the simulation i want to add a Sampler that puts in breakpoints on the switching points. 
![breakpoints](https://user-images.githubusercontent.com/57145245/107016666-82b00700-679e-11eb-9c7d-37bbb00ca060.PNG)

In a method inside the MySwitch class i am just looking on the current simulation point handed over by the "BeforeLoad" event.
![BeforeLoad](https://user-images.githubusercontent.com/57145245/107016700-8cd20580-679e-11eb-8b89-4862280674c3.PNG)
![MySwitch_temperature](https://user-images.githubusercontent.com/57145245/107016727-92c7e680-679e-11eb-8555-0745ba819680.PNG)

The problem is i am getting an exception during the creation of the behaviors because the scanassembly method does not find any behaviors for MySwitch.
![ScanAssemblys_exception](https://user-images.githubusercontent.com/57145245/107016906-cc005680-679e-11eb-8258-cb2c7af57946.PNG)

This idea already worked when i didnt pack it inside an own class but if i have more than one switch my code is getting kinda messy therefore i wanted to put it into a class. I think the problem really is only with the behaviors since the ScanAssembly method does not recognize the behaviors from Subcricuit as behaviors from MySwitch (that inherits from Subcircuit)

Thanks again for your time and advice!

**Reply by svenboulanger** [reply]
Have you perhaps considered using `VoltageSwitch` in combination with a `VoltageSource` using a `Pulse`-waveform? The pulse waveform also adds breakpoints when it changes state, and a `VoltageSwitch` also has an 'on' and an 'off' resistance.

**Reply by jaminSch** [reply]
Not really deeply, i think i misinterpreted the behavior of the VoltageSwitch but this might actually be what i need... I will look into it deeper, thank you!
